// @mikro-orm/core/metadata/MetadataStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataStorage = void 0;
const typings_1 = require("../typings");
const Utils_1 = require("../utils/Utils");
const errors_1 = require("../errors");
const EntityHelper_1 = require("../entity/EntityHelper");
class MetadataStorage {
    constructor(metadata = {}) {
        this.metadata = Utils_1.Utils.copy(metadata, false);
    }
    static getMetadata(entity, path) {
        const key = entity && path ? entity + '-' + Utils_1.Utils.hash(path) : null;
        if (key && !MetadataStorage.metadata[key]) {
            MetadataStorage.metadata[key] = new typings_1.EntityMetadata({ className: entity, path });
        }
        if (key) {
            return MetadataStorage.metadata[key];
        }
        return MetadataStorage.metadata;
    }
    static isKnownEntity(name) {
        return !!Object.values(this.metadata).find(meta => meta.className === name);
    }
    static getMetadataFromDecorator(target) {
        const path = Utils_1.Utils.lookupPathFromDecorator(target.name);
        const meta = MetadataStorage.getMetadata(target.name, path);
        Object.defineProperty(target, '__path', { value: path, writable: true });
        return meta;
    }
    static getSubscriberMetadata() {
        return MetadataStorage.subscribers;
    }
    static init() {
        return new MetadataStorage(MetadataStorage.metadata);
    }
    static clear() {
        Object.keys(this.metadata).forEach(k => delete this.metadata[k]);
        Object.keys(this.subscribers).forEach(k => delete this.subscribers[k]);
    }
    getAll() {
        return this.metadata;
    }
    getByDiscriminatorColumn(meta, data) {
        const value = data[meta.root.discriminatorColumn];
        if (!value) {
            return undefined;
        }
        const type = meta.root.discriminatorMap[value];
        return this.metadata[type];
    }
    get(entity, init = false, validate = true) {
        if (validate && !init && !this.has(entity)) {
            throw errors_1.MetadataError.missingMetadata(entity);
        }
        if (init && !this.has(entity)) {
            this.metadata[entity] = new typings_1.EntityMetadata();
        }
        return this.metadata[entity];
    }
    find(entity) {
        return this.metadata[entity];
    }
    has(entity) {
        return entity in this.metadata;
    }
    set(entity, meta) {
        return this.metadata[entity] = meta;
    }
    reset(entity) {
        delete this.metadata[entity];
    }
    decorate(em) {
        Object.values(this.metadata)
            .filter(meta => meta.prototype && !meta.prototype.__meta)
            .forEach(meta => EntityHelper_1.EntityHelper.decorate(meta, em));
    }
}
exports.MetadataStorage = MetadataStorage;
MetadataStorage.metadata = Utils_1.Utils.getGlobalStorage('metadata');
MetadataStorage.subscribers = Utils_1.Utils.getGlobalStorage('subscribers');
