// @mikro-orm/core/types/DateType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateType = void 0;
const Type_1 = require("./Type");
const errors_1 = require("../errors");
class DateType extends Type_1.Type {
    convertToDatabaseValue(value, platform) {
        if (value instanceof Date) {
            return value.toISOString().substr(0, 10);
        }
        if (!value || value.toString().match(/^\d{4}-\d{2}-\d{2}$/)) {
            return value;
        }
        throw errors_1.ValidationError.invalidType(DateType, value, 'JS');
    }
    convertToJSValue(value, platform) {
        if (!value || value instanceof Date) {
            return value;
        }
        const date = new Date(value);
        if (date.toString() === 'Invalid Date') {
            throw errors_1.ValidationError.invalidType(DateType, value, 'database');
        }
        return date;
    }
    compareAsType() {
        return 'string';
    }
    ensureComparable() {
        return false;
    }
    getColumnType(prop, platform) {
        return platform.getDateTypeDeclarationSQL(prop.length);
    }
    toJSON(value, platform) {
        return this.convertToDatabaseValue(value, platform);
    }
}
exports.DateType = DateType;
