// @mikro-orm/core/types/Type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Type = void 0;
const util_1 = require("util");
class Type {
    /**
     * Converts a value from its JS representation to its database representation of this type.
     */
    convertToDatabaseValue(value, platform, context) {
        return value;
    }
    /**
     * Converts a value from its database representation to its JS representation of this type.
     */
    convertToJSValue(value, platform) {
        return value;
    }
    /**
     * How should the raw database values be compared? Used in `EntityComparator`.
     * Possible values: string | number | boolean | date | any | buffer | array
     */
    compareAsType() {
        return 'any';
    }
    /**
     * When a value is hydrated, we convert it back to the database value to ensure comparability,
     * as often the raw database response is not the same as the `convertToDatabaseValue` result.
     * This allows to disable the additional conversion in case you know it is not needed.
     */
    ensureComparable(meta, prop) {
        return true;
    }
    /**
     * Converts a value from its JS representation to its serialized JSON form of this type.
     * By default uses the runtime value.
     */
    toJSON(value, platform) {
        return value;
    }
    /**
     * Gets the SQL declaration snippet for a field of this type.
     */
    getColumnType(prop, platform) {
        return prop.columnTypes?.[0] ?? platform.getTextTypeDeclarationSQL(prop);
    }
    static getType(cls) {
        const key = cls.name;
        if (!Type.types.has(key)) {
            Type.types.set(key, new cls());
        }
        return Type.types.get(key);
    }
    /**
     * Checks whether the argument is instance of `Type`.
     */
    static isMappedType(data) {
        return !!data?.__mappedType;
    }
    /**
     * Checks whether the argument is instance of custom `Type` class provided by the user.
     */
    static isCustomType(data) {
        return false;
    }
    [util_1.inspect.custom](depth) {
        const object = { ...this };
        const hidden = ['prop', 'platform', 'meta'];
        hidden.forEach(k => delete object[k]);
        const ret = (0, util_1.inspect)(object, { depth });
        const name = this.constructor.name;
        /* istanbul ignore next */
        return ret === '[Object]' ? `[${name}]` : name + ' ' + ret;
    }
}
exports.Type = Type;
Type.types = new Map();
Object.defineProperties(Type.prototype, {
    __mappedType: { value: true, enumerable: false, writable: false },
});
