// @mikro-orm/knex/SqlEntityRepository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqlEntityRepository = void 0;
const core_1 = require("@mikro-orm/core");
class SqlEntityRepository extends core_1.EntityRepository {
    constructor(_em, entityName) {
        super(_em, entityName);
        this._em = _em;
        this.entityName = entityName;
    }
    /**
     * Creates a QueryBuilder instance
     */
    createQueryBuilder(alias) {
        return this.getEntityManager().createQueryBuilder(this.entityName, alias);
    }
    /**
     * Shortcut for `createQueryBuilder()`
     */
    qb(alias) {
        return this.createQueryBuilder(alias);
    }
    /**
     * Returns configured knex instance.
     */
    getKnex(type) {
        return this.getEntityManager().getConnection(type).getKnex();
    }
    /**
     * @inheritDoc
     */
    getEntityManager() {
        return this._em;
    }
    /**
     * @inheritDoc
     */
    get em() {
        return this._em;
    }
}
exports.SqlEntityRepository = SqlEntityRepository;
