// @mikro-orm/knex/query/CriteriaNodeFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CriteriaNodeFactory = void 0;
const core_1 = require("@mikro-orm/core");
const ObjectCriteriaNode_1 = require("./ObjectCriteriaNode");
const ArrayCriteriaNode_1 = require("./ArrayCriteriaNode");
const ScalarCriteriaNode_1 = require("./ScalarCriteriaNode");
const CriteriaNode_1 = require("./CriteriaNode");
/**
 * @internal
 */
class CriteriaNodeFactory {
    static createNode(metadata, entityName, payload, parent, key) {
        const customExpression = CriteriaNode_1.CriteriaNode.isCustomExpression(key || '');
        const scalar = core_1.Utils.isPrimaryKey(payload) || payload instanceof RegExp || payload instanceof Date || customExpression;
        if (Array.isArray(payload) && !scalar) {
            return this.createArrayNode(metadata, entityName, payload, parent, key);
        }
        if (core_1.Utils.isPlainObject(payload) && !scalar) {
            return this.createObjectNode(metadata, entityName, payload, parent, key);
        }
        return this.createScalarNode(metadata, entityName, payload, parent, key);
    }
    static createScalarNode(metadata, entityName, payload, parent, key) {
        const node = new ScalarCriteriaNode_1.ScalarCriteriaNode(metadata, entityName, parent, key);
        node.payload = payload;
        return node;
    }
    static createArrayNode(metadata, entityName, payload, parent, key) {
        const node = new ArrayCriteriaNode_1.ArrayCriteriaNode(metadata, entityName, parent, key);
        node.payload = payload.map((item, index) => {
            const n = this.createNode(metadata, entityName, item, node);
            n.index = key === '$and' ? index : undefined; // we care about branching only for $and
            return n;
        });
        return node;
    }
    static createObjectNode(metadata, entityName, payload, parent, key) {
        const meta = metadata.find(entityName);
        const node = new ObjectCriteriaNode_1.ObjectCriteriaNode(metadata, entityName, parent, key);
        node.payload = Object.keys(payload).reduce((o, item) => {
            o[item] = this.createObjectItemNode(metadata, entityName, node, payload, item, meta);
            return o;
        }, {});
        return node;
    }
    static createObjectItemNode(metadata, entityName, node, payload, item, meta) {
        const prop = meta?.properties[item];
        const childEntity = prop && prop.reference !== core_1.ReferenceType.SCALAR ? prop.type : entityName;
        if (prop?.reference !== core_1.ReferenceType.EMBEDDED) {
            return this.createNode(metadata, childEntity, payload[item], node, item);
        }
        if (payload[item] == null) {
            const map = Object.keys(prop.embeddedProps).reduce((oo, k) => {
                oo[prop.embeddedProps[k].name] = null;
                return oo;
            }, {});
            return this.createNode(metadata, entityName, map, node, item);
        }
        // array operators can be used on embedded properties
        const allowedOperators = ['$contains', '$contained', '$overlap'];
        const operator = Object.keys(payload[item]).some(f => core_1.Utils.isOperator(f) && !allowedOperators.includes(f));
        if (operator) {
            throw core_1.ValidationError.cannotUseOperatorsInsideEmbeddables(entityName, prop.name, payload);
        }
        const map = Object.keys(payload[item]).reduce((oo, k) => {
            if (!prop.embeddedProps[k] && !allowedOperators.includes(k)) {
                throw core_1.ValidationError.invalidEmbeddableQuery(entityName, k, prop.type);
            }
            if (prop.embeddedProps[k]) {
                oo[prop.embeddedProps[k].name] = payload[item][k];
            }
            else if (typeof payload[item][k] === 'object') {
                oo[k] = JSON.stringify(payload[item][k]);
            }
            else {
                oo[k] = payload[item][k];
            }
            return oo;
        }, {});
        return this.createNode(metadata, entityName, map, node, item);
    }
}
exports.CriteriaNodeFactory = CriteriaNodeFactory;
