// @mikro-orm/knex/query/QueryBuilderHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryBuilderHelper = void 0;
const util_1 = require("util");
const core_1 = require("@mikro-orm/core");
const enums_1 = require("./enums");
/**
 * @internal
 */
class QueryBuilderHelper {
    constructor(entityName, alias, aliasMap, subQueries, knex, driver) {
        this.entityName = entityName;
        this.alias = alias;
        this.aliasMap = aliasMap;
        this.subQueries = subQueries;
        this.knex = knex;
        this.driver = driver;
        this.platform = this.driver.getPlatform();
        this.metadata = this.driver.getMetadata();
    }
    mapper(field, type = enums_1.QueryType.SELECT, value, alias) {
        if (typeof field !== 'string') {
            return field;
        }
        const isTableNameAliasRequired = this.isTableNameAliasRequired(type);
        const fields = core_1.Utils.splitPrimaryKeys(field);
        if (fields.length > 1) {
            const parts = [];
            for (const p of fields) {
                const [a, f] = this.splitField(p);
                const prop = this.getProperty(f, a);
                const fkIdx2 = prop?.fieldNames.findIndex(name => name === f) ?? -1;
                if (fkIdx2 !== -1) {
                    parts.push(this.mapper(a !== this.alias ? `${a}.${prop.fieldNames[fkIdx2]}` : prop.fieldNames[fkIdx2], type, value, alias));
                }
                else if (prop) {
                    parts.push(...prop.fieldNames.map(f => this.mapper(a !== this.alias ? `${a}.${f}` : f, type, value, alias)));
                }
                else {
                    parts.push(this.mapper(a !== this.alias ? `${a}.${f}` : f, type, value, alias));
                }
            }
            // flatten the value if we see we are expanding nested composite key
            // hackish, but cleaner solution would require quite a lot of refactoring
            if (fields.length !== parts.length && Array.isArray(value)) {
                value.forEach(row => {
                    if (Array.isArray(row)) {
                        const tmp = core_1.Utils.flatten(row);
                        row.length = 0;
                        row.push(...tmp);
                    }
                });
            }
            return this.knex.raw('(' + parts.map(part => this.knex.ref(part)).join(', ') + ')');
        }
        let ret = field;
        const customExpression = QueryBuilderHelper.isCustomExpression(field, !!alias);
        const [a, f] = this.splitField(field);
        const prop = this.getProperty(f, a);
        const fkIdx2 = prop?.fieldNames.findIndex(name => name === f) ?? -1;
        const fkIdx = fkIdx2 === -1 ? 0 : fkIdx2;
        // embeddable nested path instead of a regular property with table alias, reset alias
        if (prop?.name === a && prop.embeddedProps[f]) {
            return this.alias + '.' + prop.fieldNames[fkIdx];
        }
        const noPrefix = prop && prop.persist === false;
        if (prop?.fieldNameRaw) {
            return this.knex.raw(this.prefix(field, isTableNameAliasRequired));
        }
        if (prop?.formula) {
            const alias2 = this.knex.ref(a).toString();
            const aliased = this.knex.ref(prop.fieldNames[0]).toString();
            const as = alias === null ? '' : ` as ${aliased}`;
            return this.knex.raw(`${prop.formula(alias2)}${as}`);
        }
        if (prop?.hasConvertToJSValueSQL) {
            let valueSQL;
            if (prop.fieldNames.length > 1 && fkIdx !== -1) {
                const fk = prop.targetMeta.getPrimaryProps()[fkIdx];
                const prefixed = this.prefix(field, isTableNameAliasRequired, true, fkIdx);
                valueSQL = fk.customType.convertToJSValueSQL(prefixed, this.platform);
            }
            else {
                const prefixed = this.prefix(field, isTableNameAliasRequired, true);
                valueSQL = prop.customType.convertToJSValueSQL(prefixed, this.platform);
            }
            if (alias === null) {
                return this.knex.raw(valueSQL);
            }
            return this.knex.raw(`${valueSQL} as ${this.platform.quoteIdentifier(alias ?? prop.fieldNames[fkIdx])}`);
        }
        // do not wrap custom expressions
        if (!customExpression) {
            ret = this.prefix(field, false, false, fkIdx);
        }
        if (alias) {
            ret += ' as ' + alias;
        }
        if (customExpression) {
            const bindings = core_1.Utils.asArray(value).slice(0, ret.match(/\?/g)?.length ?? 0);
            return this.knex.raw(ret, bindings);
        }
        if (!isTableNameAliasRequired || this.isPrefixed(ret) || noPrefix) {
            return ret;
        }
        return this.alias + '.' + ret;
    }
    processData(data, convertCustomTypes, multi = false) {
        if (Array.isArray(data)) {
            return data.map(d => this.processData(d, convertCustomTypes, true));
        }
        const meta = this.metadata.find(this.entityName);
        data = this.mapData(data, meta?.properties, convertCustomTypes);
        if (!core_1.Utils.hasObjectKeys(data) && meta && multi) {
            /* istanbul ignore next */
            data[meta.primaryKeys[0]] = this.platform.usesDefaultKeyword() ? this.knex.raw('default') : undefined;
        }
        return data;
    }
    joinOneToReference(prop, ownerAlias, alias, type, cond = {}) {
        const prop2 = prop.targetMeta.properties[prop.mappedBy || prop.inversedBy];
        const table = this.getTableName(prop.type);
        const schema = prop.targetMeta?.schema === '*' ? '*' : this.driver.getSchemaName(prop.targetMeta);
        const joinColumns = prop.owner ? prop.referencedColumnNames : prop2.joinColumns;
        const inverseJoinColumns = prop.referencedColumnNames;
        const primaryKeys = prop.owner ? prop.joinColumns : prop2.referencedColumnNames;
        return {
            prop, type, cond, ownerAlias, alias, table, schema,
            joinColumns, inverseJoinColumns, primaryKeys,
        };
    }
    joinManyToOneReference(prop, ownerAlias, alias, type, cond = {}) {
        return {
            prop, type, cond, ownerAlias, alias,
            table: this.getTableName(prop.type),
            schema: prop.targetMeta?.schema === '*' ? '*' : this.driver.getSchemaName(prop.targetMeta),
            joinColumns: prop.referencedColumnNames,
            primaryKeys: prop.fieldNames,
        };
    }
    joinManyToManyReference(prop, ownerAlias, alias, pivotAlias, type, cond, path) {
        const pivotMeta = this.metadata.find(prop.pivotEntity);
        const ret = {
            [`${ownerAlias}.${prop.name}#${pivotAlias}`]: {
                prop, type, ownerAlias,
                alias: pivotAlias,
                inverseAlias: alias,
                joinColumns: prop.joinColumns,
                inverseJoinColumns: prop.inverseJoinColumns,
                primaryKeys: prop.referencedColumnNames,
                cond: {},
                table: pivotMeta.tableName,
                schema: prop.targetMeta?.schema === '*' ? '*' : this.driver.getSchemaName(pivotMeta),
                path: path.endsWith('[pivot]') ? path : `${path}[pivot]`,
            },
        };
        if (type === 'pivotJoin') {
            return ret;
        }
        const prop2 = prop.owner ? pivotMeta.relations[1] : pivotMeta.relations[0];
        ret[`${pivotAlias}.${prop2.name}#${alias}`] = this.joinManyToOneReference(prop2, pivotAlias, alias, type, cond);
        ret[`${pivotAlias}.${prop2.name}#${alias}`].path = path;
        return ret;
    }
    joinPivotTable(field, prop, ownerAlias, alias, type, cond = {}) {
        const pivotMeta = this.metadata.find(field);
        const prop2 = pivotMeta.relations[0] === prop ? pivotMeta.relations[1] : pivotMeta.relations[0];
        return {
            prop, type, cond, ownerAlias, alias,
            table: pivotMeta.collection,
            schema: pivotMeta.schema,
            joinColumns: prop.joinColumns,
            inverseJoinColumns: prop2.joinColumns,
            primaryKeys: prop.referencedColumnNames,
        };
    }
    processJoins(qb, joins, schema) {
        Object.values(joins).forEach(join => {
            let table = `${join.table} as ${join.alias}`;
            const method = join.type === 'innerJoin' ? 'inner join' : 'left join';
            schema = join.schema && join.schema !== '*' ? join.schema : schema;
            if (schema) {
                table = `${schema}.${table}`;
            }
            const conditions = [];
            const params = [];
            join.primaryKeys.forEach((primaryKey, idx) => {
                const right = `${join.alias}.${join.joinColumns[idx]}`;
                if (join.prop.formula) {
                    const left = join.prop.formula(join.ownerAlias);
                    conditions.push(`${left} = ${this.knex.ref(right)}`);
                    return;
                }
                const left = `${join.ownerAlias}.${primaryKey}`;
                conditions.push(`${this.knex.ref(left)} = ${this.knex.ref(right)}`);
            });
            Object.keys(join.cond).forEach(key => {
                conditions.push(this.processJoinClause(key, join.cond[key], params));
            });
            return qb.joinRaw(`${method} ${this.knex.ref(table)} on ${conditions.join(' and ')}`, params);
        });
    }
    processJoinClause(key, value, params, operator = '$eq') {
        if (core_1.Utils.isGroupOperator(key) && Array.isArray(value)) {
            const parts = value.map(sub => {
                return this.wrapQueryGroup(Object.keys(sub).map(k => this.processJoinClause(k, sub[k], params)));
            });
            return this.wrapQueryGroup(parts, key);
        }
        if (this.isSimpleRegExp(value)) {
            params.push(this.getRegExpParam(value));
            return `${this.knex.ref(this.mapper(key))} like ?`;
        }
        if (value instanceof RegExp) {
            value = this.platform.getRegExpValue(value);
        }
        if (core_1.Utils.isOperator(key, false) && core_1.Utils.isPlainObject(value)) {
            const parts = Object.keys(value).map(k => this.processJoinClause(k, value[k], params, key));
            return key === '$not' ? `not ${this.wrapQueryGroup(parts)}` : this.wrapQueryGroup(parts);
        }
        if (core_1.Utils.isPlainObject(value) && Object.keys(value).every(k => core_1.Utils.isOperator(k, false))) {
            const parts = Object.keys(value).map(op => this.processJoinClause(key, value[op], params, op));
            return this.wrapQueryGroup(parts);
        }
        operator = operator === '$not' ? '$eq' : operator;
        if (value === null) {
            return `${this.knex.ref(this.mapper(key))} is ${operator === '$ne' ? 'not ' : ''}null`;
        }
        if (operator === '$fulltext') {
            const [fromAlias, fromField] = this.splitField(key);
            const property = this.getProperty(fromField, fromAlias);
            const query = this.knex.raw(this.platform.getFullTextWhereClause(property), {
                column: this.mapper(key),
                query: this.knex.raw('?'),
            }).toSQL().toNative();
            params.push(value);
            return query.sql;
        }
        const replacement = this.getOperatorReplacement(operator, { [operator]: value });
        if (['$in', '$nin'].includes(operator) && Array.isArray(value)) {
            params.push(...value);
            return `${this.knex.ref(this.mapper(key))} ${replacement} (${value.map(() => '?').join(', ')})`;
        }
        if (operator === '$exists') {
            value = null;
        }
        if (QueryBuilderHelper.isCustomExpression(key)) {
            // unwind parameters when ? found in field name
            const count = key.concat('?').match(/\?/g).length - 1;
            const values = core_1.Utils.asArray(value);
            const params1 = values.slice(0, count).map((c) => core_1.Utils.isObject(c) ? JSON.stringify(c) : c);
            const params2 = values.slice(count);
            const k = this.mapper(key, enums_1.QueryType.SELECT, params1);
            const { sql, bindings } = k.toSQL().toNative();
            if (params2.length > 0) {
                params.push(...bindings);
                params.push(...params2);
                return sql + ' = ?';
            }
            params.push(...bindings);
            return sql;
        }
        const sql = this.mapper(key);
        if (value !== null) {
            params.push(value);
        }
        return `${this.knex.ref(sql)} ${replacement} ${value === null ? 'null' : '?'}`;
    }
    wrapQueryGroup(parts, operator = '$and') {
        if (parts.length === 1) {
            return parts[0];
        }
        return `(${parts.join(` ${core_1.GroupOperator[operator]} `)})`;
    }
    mapJoinColumns(type, join) {
        if (join.prop && join.prop.reference === core_1.ReferenceType.ONE_TO_ONE && !join.prop.owner) {
            return join.prop.fieldNames.map((fieldName, idx) => {
                return this.mapper(`${join.alias}.${join.inverseJoinColumns[idx]}`, type, undefined, fieldName);
            });
        }
        return [
            ...join.joinColumns.map(col => this.mapper(`${join.alias}.${col}`, type, undefined, `fk__${col}`)),
            ...join.inverseJoinColumns.map(col => this.mapper(`${join.alias}.${col}`, type, undefined, `fk__${col}`)),
        ];
    }
    isOneToOneInverse(field) {
        const meta = this.metadata.find(this.entityName);
        const prop = meta.properties[field];
        return prop && prop.reference === core_1.ReferenceType.ONE_TO_ONE && !prop.owner;
    }
    getTableName(entityName) {
        const meta = this.metadata.find(entityName);
        return meta ? meta.collection : entityName;
    }
    /**
     * Checks whether the RE can be rewritten to simple LIKE query
     */
    isSimpleRegExp(re) {
        if (!(re instanceof RegExp)) {
            return false;
        }
        if (re.flags.includes('i')) {
            return false;
        }
        // when including the opening bracket/paren we consider it complex
        return !re.source.match(/[{[(]/);
    }
    getRegExpParam(re) {
        const value = re.source
            .replace(/\.\*/g, '%') // .* -> %
            .replace(/\./g, '_') // .  -> _
            .replace(/\\_/g, '.') // \. -> .
            .replace(/^\^/g, '') // remove ^ from start
            .replace(/\$$/g, ''); // remove $ from end
        if (re.source.startsWith('^') && re.source.endsWith('$')) {
            return value;
        }
        if (re.source.startsWith('^')) {
            return value + '%';
        }
        if (re.source.endsWith('$')) {
            return '%' + value;
        }
        return `%${value}%`;
    }
    appendOnConflictClause(type, onConflict, qb) {
        onConflict.forEach(item => {
            const sub = item.fields.length > 0 ? qb.onConflict(item.fields) : qb.onConflict();
            core_1.Utils.runIfNotEmpty(() => sub.ignore(), item.ignore);
            core_1.Utils.runIfNotEmpty(() => {
                let mergeParam = item.merge;
                if (core_1.Utils.isObject(item.merge)) {
                    mergeParam = {};
                    Object.keys(item.merge).forEach(key => {
                        const k = this.mapper(key, type);
                        mergeParam[k] = item.merge[key];
                    });
                }
                if (Array.isArray(item.merge)) {
                    mergeParam = item.merge.map(key => this.mapper(key, type));
                }
                const sub2 = sub.merge(mergeParam);
                core_1.Utils.runIfNotEmpty(() => this.appendQueryCondition(type, item.where, sub2), item.where);
            }, 'merge' in item);
        });
    }
    appendQueryCondition(type, cond, qb, operator, method = 'where') {
        const m = operator === '$or' ? 'orWhere' : 'andWhere';
        Object.keys(cond).forEach(k => {
            if (k === '$and' || k === '$or') {
                if (operator) {
                    return qb[m](inner => this.appendGroupCondition(type, inner, k, method, cond[k]));
                }
                return this.appendGroupCondition(type, qb, k, method, cond[k]);
            }
            if (k === '$not') {
                const m = operator === '$or' ? 'orWhereNot' : 'whereNot';
                return qb[m](inner => this.appendQueryCondition(type, cond[k], inner));
            }
            this.appendQuerySubCondition(qb, type, method, cond, k, operator);
        });
    }
    appendQuerySubCondition(qb, type, method, cond, key, operator) {
        const m = operator === '$or' ? 'orWhere' : method;
        if (this.isSimpleRegExp(cond[key])) {
            return void qb[m](this.mapper(key, type), 'like', this.getRegExpParam(cond[key]));
        }
        if (core_1.Utils.isPlainObject(cond[key]) || cond[key] instanceof RegExp) {
            return this.processObjectSubCondition(cond, key, qb, method, m, type);
        }
        if (QueryBuilderHelper.isCustomExpression(key)) {
            return this.processCustomExpression(qb, m, key, cond, type);
        }
        const op = cond[key] === null ? 'is' : '=';
        if (this.subQueries[key]) {
            return void qb[m](this.knex.raw(`(${this.subQueries[key]})`), op, cond[key]);
        }
        qb[m](this.mapper(key, type, cond[key], null), op, cond[key]);
    }
    processCustomExpression(clause, m, key, cond, type) {
        // unwind parameters when ? found in field name
        const count = key.concat('?').match(/\?/g).length - 1;
        const value = core_1.Utils.asArray(cond[key]);
        const params1 = value.slice(0, count).map((c) => core_1.Utils.isObject(c) ? JSON.stringify(c) : c);
        const params2 = value.slice(count);
        const k = this.mapper(key, type, params1);
        if (params2.length > 0) {
            const val = params2.length === 1 && params2[0] === null ? null : this.knex.raw('?', params2);
            return void clause[m](k, val);
        }
        clause[m](k);
    }
    processObjectSubCondition(cond, key, qb, method, m, type) {
        // grouped condition for one field
        let value = cond[key];
        const size = core_1.Utils.getObjectKeysSize(value);
        if (core_1.Utils.isPlainObject(value) && size === 0) {
            return;
        }
        if (size > 1) {
            const subCondition = Object.entries(value).map(([subKey, subValue]) => ({ [key]: { [subKey]: subValue } }));
            return subCondition.forEach(sub => this.appendQueryCondition(type, sub, qb, '$and', method));
        }
        if (value instanceof RegExp) {
            value = this.platform.getRegExpValue(value);
        }
        // operators
        const op = Object.keys(core_1.QueryOperator).find(op => op in value);
        /* istanbul ignore next */
        if (!op) {
            throw new Error(`Invalid query condition: ${(0, util_1.inspect)(cond)}`);
        }
        const replacement = this.getOperatorReplacement(op, value);
        const fields = core_1.Utils.splitPrimaryKeys(key);
        if (fields.length > 1 && Array.isArray(value[op]) && !value[op].every((v) => Array.isArray(v))) {
            value[op] = this.knex.raw(`(${fields.map(() => '?').join(', ')})`, value[op]);
        }
        if (this.subQueries[key]) {
            return void qb[m](this.knex.raw(`(${this.subQueries[key]})`), replacement, value[op]);
        }
        if (op === '$fulltext') {
            const [a, f] = this.splitField(key);
            const prop = this.getProperty(f, a);
            /* istanbul ignore next */
            if (!prop) {
                throw new Error(`Cannot use $fulltext operator on ${key}, property not found`);
            }
            qb[m](this.knex.raw(this.platform.getFullTextWhereClause(prop), {
                column: this.mapper(key, type, undefined, null),
                query: value[op],
            }));
        }
        else {
            const mappedKey = this.mapper(key, type, value[op], null);
            qb[m](mappedKey, replacement, value[op]);
        }
    }
    getOperatorReplacement(op, value) {
        let replacement = core_1.QueryOperator[op];
        if (op === '$exists') {
            replacement = value[op] ? 'is not' : 'is';
            value[op] = null;
        }
        if (value[op] === null && ['$eq', '$ne'].includes(op)) {
            replacement = op === '$eq' ? 'is' : 'is not';
        }
        if (op === '$re') {
            replacement = this.platform.getRegExpOperator(value[op], value.$flags);
        }
        return replacement;
    }
    getQueryOrder(type, orderBy, populate) {
        if (Array.isArray(orderBy)) {
            return orderBy
                .map(o => this.getQueryOrder(type, o, populate))
                .filter(o => o)
                .join(', ');
        }
        return this.getQueryOrderFromObject(type, orderBy, populate);
    }
    getQueryOrderFromObject(type, orderBy, populate) {
        const ret = [];
        Object.keys(orderBy).forEach(k => {
            const direction = orderBy[k];
            const order = core_1.Utils.isNumber(direction) ? core_1.QueryOrderNumeric[direction] : direction;
            if (QueryBuilderHelper.isCustomExpression(k)) {
                ret.push(`${k} ${order.toLowerCase()}`);
                return;
            }
            core_1.Utils.splitPrimaryKeys(k).forEach(f => {
                // eslint-disable-next-line prefer-const
                let [alias, field] = this.splitField(f, true);
                alias = populate[alias] || alias;
                const prop = this.getProperty(field, alias);
                const noPrefix = (prop && prop.persist === false && !prop.formula) || QueryBuilderHelper.isCustomExpression(f);
                const column = this.mapper(noPrefix ? field : `${alias}.${field}`, type, undefined, null);
                /* istanbul ignore next */
                const rawColumn = core_1.Utils.isString(column) ? column.split('.').map(e => this.knex.ref(e)).join('.') : column;
                const customOrder = prop?.customOrder;
                const colPart = customOrder
                    ? this.platform.generateCustomOrder(rawColumn, customOrder)
                    : rawColumn;
                ret.push(`${colPart} ${order.toLowerCase()}`);
            });
        });
        return ret.join(', ');
    }
    finalize(type, qb, meta, data, returning) {
        if (!meta || !data || !this.platform.usesReturningStatement()) {
            return;
        }
        // always respect explicit returning hint
        if (returning && returning.length > 0) {
            qb.returning(returning.map(field => this.mapper(field, type)));
            return;
        }
        if (type === enums_1.QueryType.INSERT) {
            const returningProps = meta.hydrateProps
                .filter(prop => prop.returning || (prop.persist !== false && ((prop.primary && prop.autoincrement) || prop.defaultRaw)))
                .filter(prop => !(prop.name in data));
            if (returningProps.length > 0) {
                qb.returning(core_1.Utils.flatten(returningProps.map(prop => prop.fieldNames)));
            }
        }
    }
    splitField(field, greedyAlias = false) {
        const parts = field.split('.');
        if (parts.length === 1) {
            return [this.alias, parts[0]];
        }
        if (greedyAlias) {
            const fromField = parts.pop();
            const fromAlias = parts.join('.');
            return [fromAlias, fromField];
        }
        const fromAlias = parts.shift();
        const fromField = parts.join('.');
        return [fromAlias, fromField];
    }
    getLockSQL(qb, lockMode, lockTables = []) {
        const meta = this.metadata.find(this.entityName);
        if (lockMode === core_1.LockMode.OPTIMISTIC && meta && !meta.versionProperty) {
            throw core_1.OptimisticLockError.lockFailed(this.entityName);
        }
        switch (lockMode) {
            case core_1.LockMode.PESSIMISTIC_READ: return void qb.forShare(...lockTables);
            case core_1.LockMode.PESSIMISTIC_WRITE: return void qb.forUpdate(...lockTables);
            case core_1.LockMode.PESSIMISTIC_PARTIAL_WRITE: return void qb.forUpdate(...lockTables).skipLocked();
            case core_1.LockMode.PESSIMISTIC_WRITE_OR_FAIL: return void qb.forUpdate(...lockTables).noWait();
            case core_1.LockMode.PESSIMISTIC_PARTIAL_READ: return void qb.forShare(...lockTables).skipLocked();
            case core_1.LockMode.PESSIMISTIC_READ_OR_FAIL: return void qb.forShare(...lockTables).noWait();
        }
    }
    updateVersionProperty(qb, data) {
        const meta = this.metadata.find(this.entityName);
        if (!meta?.versionProperty || meta.versionProperty in data) {
            return;
        }
        const versionProperty = meta.properties[meta.versionProperty];
        let sql = this.platform.quoteIdentifier(versionProperty.fieldNames[0]) + ' + 1';
        if (versionProperty.type.toLowerCase() === 'date') {
            sql = this.platform.getCurrentTimestampSQL(versionProperty.length);
        }
        qb.update(versionProperty.fieldNames[0], this.knex.raw(sql));
    }
    static isCustomExpression(field, hasAlias = false) {
        // if we do not have alias, we don't consider spaces as custom expressions
        const re = hasAlias ? /[ ?<>=()'"`:]|^\d/ : /[?<>=()'"`:]|^\d/;
        return !!field.match(re) && !field.match(/;|--/);
    }
    prefix(field, always = false, quote = false, idx) {
        let ret;
        if (!this.isPrefixed(field)) {
            const alias = always ? (quote ? this.alias : this.platform.quoteIdentifier(this.alias)) + '.' : '';
            const fieldName = this.fieldName(field, this.alias, always, idx);
            if (QueryBuilderHelper.isCustomExpression(fieldName)) {
                return fieldName;
            }
            ret = alias + fieldName;
        }
        else {
            const [a, ...rest] = field.split('.');
            const f = rest.join('.');
            ret = a + '.' + this.fieldName(f, a, always, idx);
        }
        if (quote) {
            return this.platform.quoteIdentifier(ret);
        }
        return ret;
    }
    appendGroupCondition(type, qb, operator, method, subCondition) {
        // single sub-condition can be ignored to reduce nesting of parens
        if (subCondition.length === 1 || operator === '$and') {
            return subCondition.forEach(sub => this.appendQueryCondition(type, sub, qb, undefined, method));
        }
        qb[method](outer => subCondition.forEach(sub => {
            // skip nesting parens if the value is simple = scalar or object without operators or with only single key, being the operator
            const keys = Object.keys(sub);
            const val = sub[keys[0]];
            const simple = !core_1.Utils.isPlainObject(val) || core_1.Utils.getObjectKeysSize(val) === 1 || Object.keys(val).every(k => !core_1.Utils.isOperator(k));
            if (keys.length === 1 && simple) {
                return this.appendQueryCondition(type, sub, outer, operator);
            }
            outer.orWhere(inner => this.appendQueryCondition(type, sub, inner));
        }));
    }
    isPrefixed(field) {
        return !!field.match(/[\w`"[\]]+\./);
    }
    fieldName(field, alias, always, idx = 0) {
        const prop = this.getProperty(field, alias);
        if (!prop) {
            return field;
        }
        if (prop.fieldNameRaw) {
            if (!always) {
                return prop.fieldNameRaw
                    .replace(/\[::alias::]\.?/g, '')
                    .replace(this.platform.quoteIdentifier('') + '.', '');
            }
            if (alias) {
                return prop.fieldNameRaw.replace(/\[::alias::]/g, alias);
            }
            /* istanbul ignore next */
            return prop.fieldNameRaw;
        }
        /* istanbul ignore next */
        return prop.fieldNames?.[idx] ?? field;
    }
    getProperty(field, alias) {
        const entityName = this.aliasMap[alias]?.entityName || this.entityName;
        const meta = this.metadata.find(entityName);
        // check if `alias` is not matching an embedded property name instead of alias, e.g. `address.city`
        if (alias && meta) {
            const prop = meta.properties[alias];
            if (prop?.reference === core_1.ReferenceType.EMBEDDED) {
                // we want to select the full object property so hydration works as expected
                if (prop.object) {
                    return prop;
                }
                const parts = field.split('.');
                const nest = (p) => parts.length > 0 ? nest(p.embeddedProps[parts.shift()]) : p;
                return nest(prop);
            }
        }
        if (meta) {
            if (meta.properties[field]) {
                return meta.properties[field];
            }
            return meta.relations.find(prop => prop.fieldNames?.some(name => field === name));
        }
        return undefined;
    }
    isTableNameAliasRequired(type) {
        return [enums_1.QueryType.SELECT, enums_1.QueryType.COUNT].includes(type ?? enums_1.QueryType.SELECT);
    }
    mapData(data, properties, convertCustomTypes) {
        if (!properties) {
            return data;
        }
        data = Object.assign({}, data); // copy first
        Object.keys(data).forEach(k => {
            const prop = properties[k];
            if (!prop) {
                return;
            }
            if (prop.embeddedProps && !prop.object) {
                const copy = data[k];
                delete data[k];
                Object.assign(data, this.mapData(copy, prop.embeddedProps, convertCustomTypes));
                return;
            }
            if (prop.joinColumns && Array.isArray(data[k])) {
                const copy = core_1.Utils.flatten(data[k]);
                delete data[k];
                prop.joinColumns.forEach((joinColumn, idx) => data[joinColumn] = copy[idx]);
                return;
            }
            if (prop.customType && convertCustomTypes && !this.platform.isRaw(data[k])) {
                data[k] = prop.customType.convertToDatabaseValue(data[k], this.platform, { fromQuery: true, key: k, mode: 'query-data' });
            }
            if (prop.hasConvertToDatabaseValueSQL && !this.platform.isRaw(data[k])) {
                const quoted = this.platform.quoteValue(data[k]);
                const sql = prop.customType.convertToDatabaseValueSQL(quoted, this.platform);
                data[k] = this.knex.raw(sql.replace(/\?/g, '\\?'));
            }
            if (!prop.customType && (Array.isArray(data[k]) || core_1.Utils.isPlainObject(data[k]))) {
                data[k] = JSON.stringify(data[k]);
            }
            if (prop.fieldNames) {
                core_1.Utils.renameKey(data, k, prop.fieldNames[0]);
            }
        });
        return data;
    }
}
exports.QueryBuilderHelper = QueryBuilderHelper;
