// @mikro-orm/core/utils/AbstractSchemaGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractSchemaGenerator = void 0;
const CommitOrderCalculator_1 = require("../unit-of-work/CommitOrderCalculator");
const EntityManager_1 = require("../EntityManager");
class AbstractSchemaGenerator {
    constructor(em) {
        this.em = em instanceof EntityManager_1.EntityManager ? em : undefined;
        this.driver = em instanceof EntityManager_1.EntityManager ? em.getDriver() : em;
        this.config = this.driver.config;
        this.metadata = this.driver.getMetadata();
        this.platform = this.driver.getPlatform();
        this.connection = this.driver.getConnection();
    }
    /** @deprecated use `dropSchema` and `createSchema` commands respectively */
    async generate() {
        this.notImplemented();
    }
    async createSchema() {
        this.notImplemented();
    }
    /**
     * Returns true if the database was created.
     */
    async ensureDatabase() {
        this.notImplemented();
    }
    async refreshDatabase() {
        await this.ensureDatabase();
        await this.dropSchema();
        await this.createSchema();
    }
    async clearDatabase(options) {
        for (const meta of this.getOrderedMetadata(options?.schema).reverse()) {
            await this.driver.nativeDelete(meta.className, {}, options);
        }
        if (this.em) {
            const allowGlobalContext = this.config.get('allowGlobalContext');
            this.config.set('allowGlobalContext', true);
            this.em.clear();
            this.config.set('allowGlobalContext', allowGlobalContext);
        }
    }
    async getCreateSchemaSQL() {
        this.notImplemented();
    }
    async dropSchema() {
        this.notImplemented();
    }
    async getDropSchemaSQL() {
        this.notImplemented();
    }
    async updateSchema() {
        this.notImplemented();
    }
    async getUpdateSchemaSQL() {
        this.notImplemented();
    }
    async getUpdateSchemaMigrationSQL() {
        this.notImplemented();
    }
    /**
     * creates new database and connects to it
     */
    async createDatabase(name) {
        this.notImplemented();
    }
    async dropDatabase(name) {
        this.notImplemented();
    }
    async execute(query) {
        this.notImplemented();
    }
    async ensureIndexes() {
        this.notImplemented();
    }
    getOrderedMetadata(schema) {
        const metadata = Object.values(this.metadata.getAll()).filter(meta => {
            const isRootEntity = meta.root.className === meta.className;
            return isRootEntity && !meta.embeddable && !meta.virtual;
        });
        const calc = new CommitOrderCalculator_1.CommitOrderCalculator();
        metadata.forEach(meta => calc.addNode(meta.root.className));
        let meta = metadata.pop();
        while (meta) {
            for (const prop of meta.props) {
                calc.discoverProperty(prop, meta.root.className);
            }
            meta = metadata.pop();
        }
        return calc.sort()
            .map(cls => this.metadata.find(cls))
            .filter(meta => schema ? [schema, '*'].includes(meta.schema) : meta.schema !== '*');
    }
    notImplemented() {
        throw new Error(`This method is not supported by ${this.driver.constructor.name} driver`);
    }
}
exports.AbstractSchemaGenerator = AbstractSchemaGenerator;
