// @mikro-orm/postgresql/PostgreSqlPlatform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgreSqlPlatform = void 0;
const pg_1 = require("pg");
const core_1 = require("@mikro-orm/core");
const knex_1 = require("@mikro-orm/knex");
const PostgreSqlSchemaHelper_1 = require("./PostgreSqlSchemaHelper");
const PostgreSqlExceptionConverter_1 = require("./PostgreSqlExceptionConverter");
class PostgreSqlPlatform extends knex_1.AbstractSqlPlatform {
    constructor() {
        super(...arguments);
        this.schemaHelper = new PostgreSqlSchemaHelper_1.PostgreSqlSchemaHelper(this);
        this.exceptionConverter = new PostgreSqlExceptionConverter_1.PostgreSqlExceptionConverter();
    }
    usesReturningStatement() {
        return true;
    }
    usesCascadeStatement() {
        return true;
    }
    supportsCustomPrimaryKeyNames() {
        return true;
    }
    /**
     * Postgres will complain if we try to batch update uniquely constrained property (moving the value from one entity to another).
     * This flag will result in postponing 1:1 updates (removing them from the batched query).
     * @see https://stackoverflow.com/questions/5403437/atomic-multi-row-update-with-a-unique-constraint
     */
    allowsUniqueBatchUpdates() {
        return false;
    }
    getCurrentTimestampSQL(length) {
        return `current_timestamp(${length})`;
    }
    getDateTimeTypeDeclarationSQL(column) {
        /* istanbul ignore next */
        return 'timestamptz' + (column.length != null ? `(${column.length})` : '');
    }
    getDefaultDateTimeLength() {
        return 6; // timestamptz actually means timestamptz(6)
    }
    getTimeTypeDeclarationSQL() {
        return 'time(0)';
    }
    getIntegerTypeDeclarationSQL(column) {
        if (column.autoincrement) {
            return `serial`;
        }
        return `int`;
    }
    getBigIntTypeDeclarationSQL(column) {
        /* istanbul ignore next */
        if (column.autoincrement) {
            return `bigserial`;
        }
        return 'bigint';
    }
    getTinyIntTypeDeclarationSQL(column) {
        return 'smallint';
    }
    getUuidTypeDeclarationSQL(column) {
        return `uuid`;
    }
    getFullTextWhereClause(prop) {
        if (prop.columnTypes[0] === 'tsvector') {
            return `:column: @@ plainto_tsquery('simple', :query)`;
        }
        return `to_tsvector('simple', :column:) @@ plainto_tsquery('simple', :query)`;
    }
    supportsCreatingFullTextIndex() {
        return true;
    }
    getFullTextIndexExpression(indexName, schemaName, tableName, columns) {
        const quotedTableName = this.quoteIdentifier(schemaName ? `${schemaName}.${tableName}` : tableName);
        const quotedColumnNames = columns.map(c => this.quoteIdentifier(c.name));
        const quotedIndexName = this.quoteIdentifier(indexName);
        if (columns.length === 1 && columns[0].type === 'tsvector') {
            return `create index ${quotedIndexName} on ${quotedTableName} using gin(${quotedColumnNames[0]})`;
        }
        return `create index ${quotedIndexName} on ${quotedTableName} using gin(to_tsvector('simple', ${quotedColumnNames.join(` || ' ' || `)}))`;
    }
    getRegExpOperator(val, flags) {
        if ((val instanceof RegExp && val.flags.includes('i')) || flags?.includes('i')) {
            return '~*';
        }
        return '~';
    }
    getRegExpValue(val) {
        if (val.flags.includes('i')) {
            return { $re: val.source, $flags: val.flags };
        }
        return { $re: val.source };
    }
    isBigIntProperty(prop) {
        return super.isBigIntProperty(prop) || (['bigserial', 'int8'].includes(prop.columnTypes?.[0]));
    }
    getArrayDeclarationSQL() {
        return 'text[]';
    }
    getFloatDeclarationSQL() {
        return 'real';
    }
    getDoubleDeclarationSQL() {
        return 'double precision';
    }
    getEnumTypeDeclarationSQL(column) {
        if (column.items?.every(item => core_1.Utils.isString(item))) {
            return 'text';
        }
        return `smallint`;
    }
    supportsMultipleStatements() {
        return true;
    }
    marshallArray(values) {
        const quote = (v) => v === '' || v.match(/["{},\\]/) ? JSON.stringify(v) : v;
        return `{${values.map(v => quote('' + v)).join(',')}}`;
    }
    unmarshallArray(value) {
        if (value === '{}') {
            return [];
        }
        /* istanbul ignore next */
        return value.substring(1, value.length - 1).split(',').map(v => v === `""` ? '' : v);
    }
    getBlobDeclarationSQL() {
        return 'bytea';
    }
    getJsonDeclarationSQL() {
        return 'jsonb';
    }
    getSearchJsonPropertyKey(path, type, aliased, value) {
        const first = path.shift();
        const last = path.pop();
        const root = aliased ? (0, core_1.expr)(alias => this.quoteIdentifier(`${alias}.${first}`)) : this.quoteIdentifier(first);
        const types = {
            number: 'float8',
            boolean: 'bool',
        };
        const cast = (key) => type in types ? `(${key})::${types[type]}` : key;
        let lastOperator = '->>';
        // force `->` for operator payloads with array values
        if (core_1.Utils.isPlainObject(value) && Object.keys(value).every(key => core_1.Utils.isArrayOperator(key) && Array.isArray(value[key]))) {
            lastOperator = '->';
        }
        if (path.length === 0) {
            return cast(`${root}${lastOperator}'${last}'`);
        }
        return cast(`${root}->${path.map(a => this.quoteValue(a)).join('->')}${lastOperator}'${last}'`);
    }
    quoteIdentifier(id, quote = '"') {
        return `${quote}${id.replace('.', `${quote}.${quote}`)}${quote}`;
    }
    quoteValue(value) {
        /* istanbul ignore if */
        if (core_1.Utils.isPlainObject(value) || value?.[core_1.JsonProperty]) {
            value = JSON.stringify(value);
        }
        if (typeof value === 'string') {
            return pg_1.Client.prototype.escapeLiteral(value);
        }
        if (value instanceof Date) {
            return `'${value.toISOString()}'`;
        }
        if (ArrayBuffer.isView(value)) {
            return `E'\\\\x${value.toString('hex')}'`;
        }
        return super.quoteValue(value);
    }
    indexForeignKeys() {
        return false;
    }
    getDefaultMappedType(type) {
        const normalizedType = this.extractSimpleType(type);
        const map = {
            'int2': 'smallint',
            'smallserial': 'smallint',
            'int': 'integer',
            'int4': 'integer',
            'serial': 'integer',
            'serial4': 'integer',
            'int8': 'bigint',
            'bigserial': 'bigint',
            'serial8': 'bigint',
            'numeric': 'decimal',
            'bool': 'boolean',
            'real': 'float',
            'float4': 'float',
            'float8': 'double',
            'timestamp': 'datetime',
            'timestamptz': 'datetime',
            'bytea': 'blob',
            'jsonb': 'json',
            'character varying': 'varchar',
        };
        return super.getDefaultMappedType(map[normalizedType] ?? type);
    }
    supportsSchemas() {
        return true;
    }
    getDefaultSchemaName() {
        return 'public';
    }
    /**
     * Returns the default name of index for the given columns
     * cannot go past 64 character length for identifiers in MySQL
     */
    getIndexName(tableName, columns, type) {
        const indexName = super.getIndexName(tableName, columns, type);
        if (indexName.length > 64) {
            return `${indexName.substring(0, 56 - type.length)}_${core_1.Utils.hash(indexName, 5)}_${type}`;
        }
        return indexName;
    }
    getDefaultPrimaryName(tableName, columns) {
        const indexName = `${tableName}_pkey`;
        if (indexName.length > 64) {
            return `${indexName.substring(0, 56 - 'primary'.length)}_${core_1.Utils.hash(indexName, 5)}_primary`;
        }
        return indexName;
    }
    /**
     * @inheritDoc
     */
    castColumn(prop) {
        switch (prop?.columnTypes?.[0]) {
            case this.getUuidTypeDeclarationSQL({}): return '::text';
            case this.getBooleanTypeDeclarationSQL(): return '::int';
            case 'json': return '::jsonb';
            default: return '';
        }
    }
    /**
     * @inheritDoc
     */
    castJsonValue(prop) {
        if (prop?.columnTypes?.[0] === 'json') {
            return '::text';
        }
        return '';
    }
}
exports.PostgreSqlPlatform = PostgreSqlPlatform;
